// @crypto.com/developer-platform-client/dist/integrations/api.interfaces.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Method = exports.Status = void 0;
/**
 * Enum representing the status of a CDC AI Agent response.
 *
 * @enum {string}
 * @property {string} Success - The request was successful.
 * @property {string} Failed - The request failed.
 */
var Status;
(function (Status) {
    Status["Success"] = "Success";
    Status["Failed"] = "Failed";
})(Status = exports.Status || (exports.Status = {}));
/**
 * Enum representing the HTTP methods that can be used in API requests.
 *
 * @enum {string}
 * @property {string} GET - HTTP GET method.
 * @property {string} POST - HTTP POST method.
 * @property {string} PUT - HTTP PUT method.
 * @property {string} DELETE - HTTP DELETE method.
 *
 * @example
 * const method: Method = Method.GET;
 */
var Method;
(function (Method) {
    Method["GET"] = "GET";
    Method["POST"] = "POST";
    Method["PUT"] = "PUT";
    Method["DELETE"] = "DELETE";
})(Method = exports.Method || (exports.Method = {}));
