// @crypto.com/developer-platform-client/dist/lib/client/interfaces/transaction.interfaces.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionAction = exports.ChainName = exports.Unit = exports.BlockTagString = void 0;
var BlockTagString;
(function (BlockTagString) {
    BlockTagString["Latest"] = "latest";
    BlockTagString["Earliest"] = "earliest";
})(BlockTagString = exports.BlockTagString || (exports.BlockTagString = {}));
var Unit;
(function (Unit) {
    Unit["Ether"] = "ether";
    Unit["Gwei"] = "gwei";
})(Unit = exports.Unit || (exports.Unit = {}));
var ChainName;
(function (ChainName) {
    ChainName["CRONOS_EVM"] = "Cronos EVM Mainnet";
    ChainName["CRONOS_EVM_TESTNET"] = "Cronos EVM Testnet";
    ChainName["CRONOS_ZKEVM"] = "Cronos ZK EVM Mainnet";
    ChainName["CRONOS_ZKEVM_TESTNET"] = "Cronos ZK EVM Testnet";
})(ChainName = exports.ChainName || (exports.ChainName = {}));
var TransactionAction;
(function (TransactionAction) {
    TransactionAction["Deposit"] = "deposit";
    TransactionAction["Approve"] = "approve";
    TransactionAction["Swap"] = "swap";
})(TransactionAction = exports.TransactionAction || (exports.TransactionAction = {}));
